pkg/tests/api/azuremonitor/azuremonitor_test.go

```diff
@@ -0,0 +1,117 @@
+
+	var outgoingRequest *http.Request
+	outgoingServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		outgoingRequest = r
+		w.WriteHeader(http.StatusUnauthorized)
+	}))
+	t.Cleanup(outgoingServer.Close)
+
+	jsonData := simplejson.NewFromAny(map[string]interface{}{
+		"httpHeaderName1": "X-CUSTOM-HEADER",
+		"clientId":        "test-client-id",
+		"tenantId":        "test-tenant-id",
+		"cloudName":       "customizedazuremonitor",
+		"customizedRoutes": map[string]interface{}{
+			"Azure Monitor": map[string]interface{}{
+				"URL": outgoingServer.URL,
+			},
+		},
+	})
+	secureJSONData := map[string]string{
+		"clientSecret":     "test-client-secret",
+		"httpHeaderValue1": "custom-header-value",
+	}
+
+	uid := "azuremonitor"
+	err := testEnv.Server.HTTPServer.DataSourcesService.AddDataSource(ctx, &datasources.AddDataSourceCommand{
+		OrgId:          1,
+		Access:         datasources.DS_ACCESS_PROXY,
+		Name:           "Azure Monitor",
+		Type:           datasources.DS_AZURE_MONITOR,
+		Uid:            uid,
+		Url:            outgoingServer.URL,
+		JsonData:       jsonData,
+		SecureJsonData: secureJSONData,
+	})
+	require.NoError(t, err)
+
+	t.Run("When calling /api/ds/query should set expected headers on outgoing HTTP request", func(t *testing.T) {
+		query := simplejson.NewFromAny(map[string]interface{}{
+			"datasource": map[string]interface{}{
+				"type": "grafana-azure-monitor-datasource",
+				"uid":  uid,
+			},
+			"queryType": "Azure Monitor",
+			"azureMonitor": map[string]interface{}{
+				"resourceGroup":   "test-rg",
+				"metricNamespace": "microsoft.storage/storageaccounts",
+				"resourceName":    "testacct",
+				"timeGrain":       "auto",
+				"metricName":      "UsedCapacity",
+				"aggregation":     "Average",
+			},
+			"subscription": "test-sub",
+			"intervalMs":   30000,
+		})
+		buf1 := &bytes.Buffer{}
+		err = json.NewEncoder(buf1).Encode(dtos.MetricRequest{
+			From:    "1668078080000",
+			To:      "1668081680000",
+			Queries: []*simplejson.Json{query},
+		})
+		require.NoError(t, err)
+		u := fmt.Sprintf("http://admin:admin@%s/api/ds/query", grafanaListeningAddr)
+		// nolint:gosec
+		resp, err := http.Post(u, "application/json", buf1)
+		require.NoError(t, err)
+		require.Equal(t, http.StatusBadRequest, resp.StatusCode)
+		t.Cleanup(func() {
+			err := resp.Body.Close()
+			require.NoError(t, err)
+		})
+		_, err = io.ReadAll(resp.Body)
+		require.NoError(t, err)
+
+		require.NotNil(t, outgoingRequest)
+		require.Equal(t, "/subscriptions/test-sub/resourceGroups/test-rg/providers/microsoft.storage/storageaccounts/testacct/providers/microsoft.insights/metrics?aggregation=Average&api-version=2021-05-01&interval=PT1M&metricnames=UsedCapacity&metricnamespace=microsoft.storage%2Fstorageaccounts&timespan=2022-11-10T11%3A01%3A20Z%2F2022-11-10T12%3A01%3A20Z",
+			outgoingRequest.URL.String())
+		require.Equal(t, "custom-header-value", outgoingRequest.Header.Get("X-CUSTOM-HEADER"))
```

Comment:
This test is correct, but still a bit weird since you can't configure custom headers using the Azure Monitor config page/UI.

Don't you want to verify route headers have been added, e.g. https://github.com/grafana/grafana/blob/b4e76e75b1176901a1e79804c56e8ccaf62414b0/pkg/tsdb/azuremonitor/routes.go#L19

Or is that not possible/tricky?

Reply:
I can add an additional test to validate the Azure headers as well given that we have to use the `customizedazureroute` to overwrite the outgoing URL. 

Is it a requirement for custom headers config to be in the datasource configuration? I thought (maybe incorrectly) that this feature was specific to provisioning?

Reply:
> Is it a requirement for custom headers config to be in the datasource configuration? I thought (maybe incorrectly) that this feature was specific to provisioning?

No, but without any documentation or noticeable UI it's very hard for anyone to understand this is supported. Usually this is mainly supported for those datasources that uses the [DataSourceHttpSettings component](https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/DataSourceSettings/DataSourceHttpSettings.tsx). Since Azure is a cloud service I don't understand why custom headers needs to/should be supported, but maybe there's some users that have requested this?

Reply:
The Grafana documentation implies that this feature is supported for all datasources rather than a specific subset. There hasn't been a specific request for custom headers although support for this as a part of a custom Azure cloud has recently been developed. 

I could imagine there may be some use cases where this could be useful to users. I misinterpreted the original issue as being related specifically to custom headers. The Google Cloud Monitoring datasource is also listed as requiring this test but is in a similar situation to Azure, can you confirm if this test should be written there as well please 😊

Reply:
Also, you're correct that validating the default route headers is difficult as the request URL is also set and tricky to overwrite. This test does validate that any headers added as a part of the route are forwarded to the outgoing request which sould be sufficient.

Reply:
> The Grafana documentation implies that this feature is supported for all datasources rather than a specific subset. There hasn't been a specific request for custom headers although support for this as a part of a custom Azure cloud has recently been developed.

Fair. I guess you're referring to https://grafana.com/docs/grafana/latest/administration/provisioning/#custom-http-headers-for-data-sources.

> The Google Cloud Monitoring datasource is also listed as requiring this test but is in a similar situation to Azure, can you confirm if this test should be written there as well please

Probably a good idea given you added it for azure. Might or might not be some additional asserts for cloud monitoring that you want to verify? 

> Also, you're correct that validating the default route headers is difficult as the request URL is also set and tricky to overwrite. This test does validate that any headers added as a part of the route are forwarded to the outgoing request which sould be sufficient.

Oh right

Reply:
Thanks for the review @marefr! I'll look at implementing something similar for GCM as well 😊